syntax = "proto3";

package fp_puncher;

message HostConnect {
    string host_name = 1;
    optional string password = 2;
}

message HostResponse {
    bool succeeded = 1;
    string token = 2;
}

message HostNewClient {
    string ip = 1;
    uint32 port = 2;
}

message HostHeartbeat {
    string host_name = 1;
    string token = 2;
}

message HostExit {
    string host_name = 1;
    string token = 2;
}

message ClientConnect {
    string host_name = 1;
    optional string password = 2;
}

message ClientResponse {
    bool succeeded = 1;
    string host_token = 2;
    string ip = 3;
    uint32 port = 4;
}

message PunchOpen { }

message ConnectMessage {
    oneof msg {
        HostConnect host = 1;
        ClientConnect client = 2;
        HostHeartbeat heartbeat = 3;
        HostExit exit = 4;
    }
}

message ServerMessage {
    oneof msg {
        HostResponse host = 1;
        HostNewClient new_client = 2;
        ClientResponse client = 3;
    }
}